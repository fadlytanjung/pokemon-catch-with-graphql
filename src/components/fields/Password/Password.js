import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styles from './styles.scoped.css';

export default function Password(props) {
  const { className, info, input, inputProps, message, meta } = props;
  const [focus, setFocus] = useState(false);
  const [show, setShow] = useState(false);
  const { dirty, error, touched } = meta;
  const isRed = error &&  (dirty || touched);
  const classes = [
    styles.root,
    isRed && styles.error,
    focus && styles.focus,
    className,
  ].filter(Boolean).join(' ');
  const infoClass = isRed ? styles.error : styles.info;
  const validation = (typeof error === 'string') && error;

  const events = {
    onBlur: e => {
      setFocus(false);
      input.onBlur(e);
    },
    onFocus: e => {
      setFocus(true);
      input.onFocus(e);
    },
  };

  return (
    <>
      <label className={styles.label}>{inputProps.label}</label>
      <div className={classes}>
        <input {...input} {...inputProps} {...events} type={show ? 'text' : 'password'} />
        <button onClick={() => setShow(!show)} type="button">
          <img alt="" src={`/assets/icon_pw_${show ? 'show' : 'hide'}.svg`} />
        </button>
      </div>
      {info ?
        <small className={infoClass}>{message}</small>:
        isRed && <small className={styles.error}>{validation || message}</small>
      }
    </>
  );
}

Password.defaultProps = {
  className: '',
  info: false,
  input: {},
  inputProps: {},
  message: '',
  meta: {},
};

Password.propTypes = {
  className: PropTypes.string,
  info: PropTypes.bool,
  input: PropTypes.object,
  inputProps: PropTypes.object,
  message: PropTypes.string,
  meta: PropTypes.object,
};
