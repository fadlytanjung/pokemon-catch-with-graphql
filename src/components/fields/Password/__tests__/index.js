import React, { useState } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Password from '../Password';

jest.mock('../styles.scoped.css', () => ({
  error: 'error',
  focus: 'focus',
  info: 'info',
  root: 'root',
}));

describe('src/components/fields/Password', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Password />);
    expect(tree).toMatchSnapshot();

    const result = Password({
      ...Password.defaultProps,
      input: { value: 'test' },
      inputProps: { label: 'abc' },
    });
    const label = result.props.children[0].props.children;
    expect(label).toBe('abc');
    const button = result.props.children[1].props.children[1];
    button.props.onClick();
  });

  test('render error & focus & show password', () => {
    useState
      .mockImplementationOnce(() => [true, jest.fn()])
      .mockImplementationOnce(() => [true, jest.fn()]);
    const meta = { error: true, touched: true };
    const result = Password({ ...Password.defaultProps, meta });
    const divRoot = result.props.children[1];
    expect(divRoot.props.className).toBe('root error focus');
    expect(divRoot.props.children[0].props.type).toBe('text');
  });

  test('render info', () => {
    const result = Password({ ...Password.defaultProps, info: true, message: 'info' });
    const small = result.props.children[2];
    expect(small.props.children).toBe('info');
    expect(small.props.className).toBe('info');
  });

  test('render info error', () => {
    const meta = { error: 'error', touched: true };
    const result = Password({ ...Password.defaultProps, info: true, message: 'info', meta });
    const small = result.props.children[2];
    expect(small.props.children).toBe('info');
    expect(small.props.className).toBe('error');
  });

  test('onFocus & onBlur', () => {
    const setFocus = jest.fn();
    const input = {
      onBlur: jest.fn(),
      onFocus: jest.fn(),
    };
    useState.mockImplementationOnce(() => [true, setFocus]);
    const result = Password({ ...Password.defaultProps, input });
    result.props.children[1].props.children[0].props.onBlur('e');
    expect(setFocus).toHaveBeenNthCalledWith(1, false);
    expect(input.onBlur).toHaveBeenCalledWith('e');
    result.props.children[1].props.children[0].props.onFocus('e');
    expect(setFocus).toHaveBeenNthCalledWith(2, true);
    expect(input.onFocus).toHaveBeenCalledWith('e');
  });
});
