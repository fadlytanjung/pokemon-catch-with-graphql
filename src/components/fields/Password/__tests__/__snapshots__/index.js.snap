// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fields/Password render 1`] = `
<React.Fragment>
  <label />
  <div
    className="root"
  >
    <input
      onBlur={[Function]}
      onFocus={[Function]}
      type="password"
    />
    <button
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        src="/assets/icon_pw_hide.svg"
      />
    </button>
  </div>
</React.Fragment>
`;
